CenterForOpenScience/ember-osf-web
# [ENG-4002] log pageviews to osf:metrics

-   Ticket: [ENG-4002]
-   Feature flag: n/a

## Purpose
start saving our own pageview/usage data (under `osf:api/_/metrics/`), without disrupting our current metrics setup but with intent to eventually delete our current metrics setup.
(depends on CenterForOpenScience/osf.io#9954)


## Summary of Changes
- update `trackPage()` on the `analytics` service to:
  - gather metadata from each active route via [`RouterService.currentRoute`](https://api.emberjs.com/ember/3.26/classes/RouterService/properties/currentRoute?anchor=currentRoute) (including route-specific metadata via [`RouteInfo.metadata`](https://api.emberjs.com/ember/3.26/classes/RouteInfo/properties/metadata?anchor=metadata))
  - send a usage event to `osf:api/_/metrics/events/counted_usage/`
- implement [`buildRouteInfoMetadata()`](https://api.emberjs.com/ember/3.26/classes/Route/methods/buildRouteInfoMetadata?anchor=buildRouteInfoMetadata) on routes that have their own pageview metadata (e.g. a guid or provider)
- add `trackPage()` calls to each `application` route's `didTransition` hook, in the host app and each engine 
- remove `trackPage()` calls from all non-`application` routes that had them
- add mirage endpoint and model for `counted_usage`, and a simple `analytics.trackPage()` test that uses them
- update the [analytics page in the handbook](https://github.com/CenterForOpenScience/ember-osf-web/pull/1640/files#diff-27cfe9fecf6f24b61e3b7b6bdd13c46ad4b0c7e5e3a9b9b0d88013d015804095) accordingly


## Screenshot(s)
n/a


## Side Effects
none?


## QA Notes
recommend:
1. given a guid for a public project (or registration), go to `osf:api/_/metrics/query/node_analytics/<guid>/week/` and note `data.attributes.popular_pages.*.path`
2. navigate to a page "under" the given guid (either a page with the guid in the url, or a sub-node/file if the guid refers to a node -- with CenterForOpenScience/osf.io#9954, shouldn't matter whether legacy-ui or ember-osf-web)
3. verify that `osf:api/_/metrics/query/node_analytics/<guid>/week/` reflects the new view

additionally, every page view/transition should send a `POST osf:api/_/metrics/counted_usage/` with appropriate info -- don't worry about checking every single one unless you want to, tho.



[ENG-4002]: https://openscience.atlassian.net/browse/ENG-4002?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
